shared/lib/trace.ts

```diff
@@ -134,35 +120,73 @@ async function startTrace(
     return span;
   };
 
-  if (!isSentryEnabled) {
-    return await callback(null);
-  }
-
-  return await startSpan(request, (spanOptions) =>
+  return startSpan(request, (spanOptions) =>
     Sentry.startSpanManual(spanOptions, callback),
   );
 }
 
-async function startSpan<T>(
+function startSpan<T>(
   request: TraceRequest,
-  callback: (spanOptions: StartSpanOptions) => Promise<T>,
+  callback: (spanOptions: StartSpanOptions) => T,
 ) {
-  const { data: attributes, name, parentContext, tags } = request;
+  const { data: attributes, name, parentContext, startTime, tags } = request;
   const parentSpan = (parentContext ?? null) as Sentry.Span | null;
-  const spanOptions = { name, parentSpan, attributes };
 
-  return await Sentry.withIsolationScope(async (scope) => {
+  const spanOptions: StartSpanOptions = {
+    attributes,
+    name,
+    op: OP_DEFAULT,
+    parentSpan,
+    startTime,
+  };
+
+  return Sentry.withIsolationScope((scope) => {
     scope.setTags(tags as Record<string, Primitive>);
 
-    return await callback(spanOptions);
+    return callback(spanOptions);
   });
 }
 
+function getTraceId(request: TraceRequest) {
+  return request.id ?? ID_DEFAULT;
+}
+
 function getTraceKey(request: TraceRequest) {
-  const { id, name } = request;
+  const { name } = request;
+  const id = getTraceId(request);
+
   return [name, id].join(':');
 }
 
 function getPerformanceTimestamp(): number {
   return performance.timeOrigin + performance.now();
 }
+
+function tryCatchMaybePromise<T>(
+  tryFn: () => T,
+  catchFn: (error: unknown) => void,
+  finallyFn: () => void,
+): T | undefined {
+  let isPromise = false;
+
+  try {
+    const result = tryFn() as T;
+
+    if (result instanceof Promise) {
```

Comment:
I don't know if we want to do anything about this, but this check isn't 100%.

An example of something that behaves like a `Promise`, but isn't an `instanceof` a `Promise`.

```
function tryFn() {
  return {
    // a "thenable" object
      then: function (callback) {
          setTimeout(() => {
              const value = 123;
              callback(value);
          }, 1000);
      }
  };
}

const result = tryFn();
console.log(result instanceof Promise); // false
console.log(typeof result.then); // function
const value = await result;
console.log(value); // 123
```

I'm not sure if just checking for a `.then` function would be reliable either.

Maybe its not an issue for our code base?

Reply:
This is a good point, but since the nature of these traces is they are added manually, I think we'd identify when added if a promise is non-typical and therefore impacting the duration.